### PR TITLE
Improve remote host/owner logic

### PR DESCRIPTION
**Pull request by matthewrmshin**
Handle situation where host can be reached by SSH but not by `socket.gethostbyname`. (E.g. Host is a setting in `~/.ssh/config`, but not otherwise known by the local name server. This would have brought the suite down.)

Remove client's pointless `get_hostname` call for `host == 'localhost'`. See this [Google Groups Thread](https://groups.google.com/forum/?fromgroups#!topic/cylc/r56vn9S04mE) for detail.

On null value for host in client, set host name using `get_suite_host` (instead of `get_hostname`) so logic should be consistent with that at start up of a suite.

Move host/owner functionality into single module. Replace `USER` with `get_user()` to defer calculation at point of use (instead of at import time). Handle local owner identifiable by multiple names.

### CODE REVIEW

**Comment by hjoliver** [comment]
(conflicts)

**Comment by matthewrmshin** [comment]
(Branch re-based.)

File: lib/cylc/network/https/base_client.py [hunk 1]
```diff
@@ -138,28 +139,21 @@ def call_server_func(self, category, *func_dicts, **fargs):
...
-        if self.host is None and self.port is not None:
-            self.host = get_hostname()
         try:
             self._load_contact_info()
         except (IOError, ValueError, SuiteServiceFileError):
             raise ConnectionInfoError(self.suite)
         handle_proxies()
-        host = self.host
-        if host == "localhost":
-            host = get_hostname().split(".")[0]
```
Comment (hjoliver) [inline lib/cylc/network/https/base_client.py]: Interestingly, due to some DNS changes (presumably) this line recently broke `cylc scan` on my desktop box at work (where hostname `foo.niwa.local` is visible, but `foo` isn't).

**Comment by hjoliver** [comment]
Warning: from @trwhitcomb on the Cylc group:

> [if I remove the two lines above] using the debug mode to try to open gcylc windows from gscan s
showed ... opening a gcylc window triggers a traceback with:
>
> SSLError: hostname ‘localhost’ doesn’t match u’<my hostname>’
>
> If, instead of removing the two lines I do
> 
> if hostname == “localhost”:
>        host = get_hostname()
>
> then I’m able to run gcylc successfully from the gscan window.

**Comment by matthewrmshin** [comment]
(Branch re-based against master. Also incorporated #2371 partially in this change, so it can pass site test in my environment.)

**Comment by matthewrmshin** [comment]
Latest version:
* On suite register/run: Remove `DNS:*` from SSL certificate.
* On `localhost` value for host in client, set host name using `get_suite_host` so logic should be consistent with that at start up of a suite.

File: lib/cylc/suite_host.py [hunk 2]
```diff
@@ -147,11 +151,26 @@ def get_suite_host(self):
                 self._host_name_pref = self.get_hostname()
         return self._host_name_pref
 
+    def get_user(self):
+        """Return name of current user."""
+        return self._get_user_pwent().pw_name
```
Comment (dvalters) [inline lib/cylc/suite_host.py]: Didn't know you could access dictionary values like this. (I was mistakenly looking for a function attribute..)

Comment (matthewrmshin) [inline lib/cylc/suite_host.py]: If I am not mistaken, `pwd.struct_passwd` is a named tuple. Very handy.

Comment (dvalters) [inline lib/cylc/suite_host.py]: Oh yes. (I was looking at `remote_users`). Apparently it can be done for dictionary keys too.

**Comment by matthewrmshin** [comment]
(Branch re-based, so it can be tested with the latest code before merge.)

**Comment by trwhitcomb** [comment]
@hjoliver Thanks - I had a chance to check it out this morning.  I'm no longer receiving the certificate validation warnings I was seeing before (although the logging is an outstanding issue as @matthewrmshin noted) and I'm now able to launch gcylc from the gscan window (where it wasn't working before).  Looks good to me!

**Comment by hjoliver** [comment]
OK, with test by @trwhitcomb - this is good to merge.

**Pull request merged**